File: cpp/src/app/Application.cpp
#include "ui/app/Application.hpp"

#include <stdexcept>

namespace ui {

Window& Application::open(std::unique_ptr<Window> window) {
    if (!window) {
        throw std::invalid_argument("window must not be null");
    }
    Window& ref = *window;
    windows_.push_back(std::move(window));
    return ref;
}

Window* Application::activeWindow() const {
    if (windows_.empty()) {
        return nullptr;
    }
    return windows_.back().get();
}

}  // namespace ui

File: java/src/main/java/uimd/Application.java
package uimd;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class Application
{
    private final List<Window> windows = new ArrayList<>();

    public Window open(Window window)
    {
        if (window == null)
        {
            throw new IllegalArgumentException("window must not be null");
        }
        windows.add(window);
        return window;
    }

    public Window activeWindow()
    {
        return windows.isEmpty() ? null : windows.get(windows.size() - 1);
    }

    public List<Window> windows()
    {
        return Collections.unmodifiableList(windows);
    }
}
